### PR TITLE
ISSUE #1561: Fixes create and restore database with drush arr

### PR DESCRIPTION
**Pull request by sherakama**
Suggested fix for issue #1561

### CODE REVIEW

File: lib/Drush/Sql/SqlBase.php [hunk 1]
```diff
@@ -257,7 +257,10 @@ public function drop_or_create() {
       $this->drop($this->listTables());
     }
     else {
+      $original_spec = $this->db_spec;
       $this->createdb();
+      // Restore the original spec as create db and su() can modify the spec.
+      $this->db_spec = $original_spec;
```
Comment (weitzman) [inline lib/Drush/Sql/SqlBase.php]: I didnt investigate, but it is a bit odd to wrap createdb here. Can this dance not be done inside createdb()?

File: lib/Drush/Sql/Sqlmysql.php [hunk 2]
```diff
@@ -32,7 +32,9 @@ public function creds($hide_password = TRUE) {
 
     // Some drush commands (e.g. site-install) want to connect to the
     // server, but not the database.  Connect to the built-in database.
-    $parameters['database'] = empty($this->db_spec['database']) ? 'information_schema' : $this->db_spec['database'];
+    if (!empty($this->db_spec['database'])) {
```
Comment (weitzman) [inline lib/Drush/Sql/Sqlmysql.php]: I'm a little hesitant to remove information_schema. I think we need to specify a DB in certain cases (see early parts of site-install command). Needs investigation.

File: lib/Drush/Sql/SqlBase.php (hunk 1 above)
Comment (sherakama) [inline lib/Drush/Sql/SqlBase.php]: That might be a better approach. I'll take a look to see.

File: lib/Drush/Sql/Sqlmysql.php (hunk 2 above)
Comment (sherakama) [inline lib/Drush/Sql/Sqlmysql.php]: information_schema cannot be written to. Read is fine, but I wonder if we can protect this from defaulting to information_schema when it is a write operation.

Comment (weitzman) [inline lib/Drush/Sql/Sqlmysql.php]: In that case the caller should have provided a database. I'm not too concerned about this.

**Comment by sherakama** [comment]
Ok, moved the db_spec restore to the createdb method and added a flag for when creating a new database to avoid adding the --database parameter to the mysql statement. 

Debug output now looks like:

```
Executing: mysql --defaults-extra-file=/private/tmp/drush_OlVRJz --database=mydatabase --host=localhost --silent  < /private/tmp/drush_L8XnDW > /dev/null

Executing: mysql --defaults-extra-file=/private/tmp/drush_dgBIe0 --host=localhost --silent  < /private/tmp/drush_NnPnGb

Executing: mysql --defaults-extra-file=/private/tmp/drush_YEIDG2 --database=mydatabase --host=localhost --silent  < /tmp/drush_tmp_1440615699_55de0d13b289c/ds_jumpstart_cars.sql
```

You can see that the second mysql statement does not include the database parameter as that one has the create new db statement in it.

File: lib/Drush/Sql/SqlBase.php [hunk 3]
```diff
@@ -239,11 +239,26 @@ public function createdb_sql($dbname, $quoted = FALSE) {}
...
+
+    // Create the query.
+    $result = $this->query($sql);
+
+    // the su() command can change the database name to an empty string which
+    // can effect further operations. Restore the original db_spec to continue.
+    $this->db_spec = $original_spec;
+
+    // Return the result of the query.
+    return $result;
```
Comment (weitzman) [inline lib/Drush/Sql/SqlBase.php]: This hunk looks good to me.

File: lib/Drush/Sql/Sqlmysql.php [hunk 4]
```diff
@@ -34,6 +34,12 @@ public function creds($hide_password = TRUE) {
     // server, but not the database.  Connect to the built-in database.
     $parameters['database'] = empty($this->db_spec['database']) ? 'information_schema' : $this->db_spec['database'];
 
+    // If we are creating a new database we don't want to include the database
+    // parameter for the mysql statement.
+    if (isset($this->db_spec['new-database']) && $this->db_spec['new-database']) {
+      unset($parameters['database']);
+    }
+
```
Comment (weitzman) [inline lib/Drush/Sql/Sqlmysql.php]: Maybe creds() should always return a full array and it is up to the caller to remove 'database' if needed?

Comment (sherakama) [inline lib/Drush/Sql/Sqlmysql.php]: Outside of creating or dropping a database I don't see a need to remove the database option. Perhaps changing the key from 'new-database' to 'no-database' or something similar?

**Comment by sherakama** [comment]
Moved this to: https://github.com/drush-ops/drush/pull/1872

**Pull request closed without merging**